### PR TITLE
Keyring webview

### PR DESCRIPTION
**Pull request by ishrak-imam**
### Description

Basic implementation of creating a keyring account using the packages through webview so that we can leverage the WASM support.

## TODO:

Feature:
- [ ] Add signing transactions support

Improvements:
- [ ] Fix TS errors
- [ ] Encapsulate setCallback within AccountsContext
- [ ] Try to avoid setTimeout while html is loading in webview
- [ ] Use the dispatched actions type strings from same file in both projects
- [ ] Add TS and linting support for the litentry-keyring project

### CODE REVIEW

**Comment by geastwood** [comment]
Does this satisfy all our need in them of keyring functions?

**Comment by ishrak-imam** [comment]
> Does this satisfy all our need in them of keyring functions?

Communication between RN and webview is not straight forward, involves message passing in string format i.e. `JSON parse and stringify`. Although this implementation has much room for improvements such as automating the build process of webview html, improving the code quality etc. it supports all our current needs as in, creating internal/external accounts, importing account from seed or json file.

Also one of the other non trivial part of this approach is that webview local storage will not persist data in between app starts. So we have to store all the accounts data in RN app local store and populate the webview local storage on app start so that webview keyring has access to the previously created accounts

**Comment by geastwood** [comment]
And how's the performance?

**Comment by ahoseinian** [comment]
Maybe we can use https://www.npmjs.com/package/json-rpc-2.0 to encapsulate the communication between WebViews and RN.

**Comment by ishrak-imam** [comment]
> And how's the performance?

Better than the normal approach, i.e. the app would be usable at least.

File: src/context/AccountsContext.tsx [hunk 1]
```diff
@@ -14,67 +18,270 @@ export type Account = {
...
+    if(localStoreAccounts.length > 0) {
+      mmkvStorage.set(KEYRING_ACCOUNTS, JSON.stringify(localStoreAccounts))
+    }
+  }, [localStoreAccounts])
 
-  return <AccountsContext.Provider value={accounts}>{children}</AccountsContext.Provider>;
-}
+  useEffect(() => {
+    if(html) {
+      setTimeout(() => {
```
Comment (silva-fj) [inline src/context/AccountsContext.tsx:133]: Why do we need `setTimeout`?

File: src/screen/CreateAccount/CreateAccountScreen.tsx [hunk 2]
```diff
@@ -26,6 +26,20 @@ export function CreateAccountScreen({
 
   const theme = useTheme();
   const {currentNetwork} = React.useContext(NetworkContext);
+  const {setCallback, createAccount, addAccount} = React.useContext(AccountsContext)
+
+  setCallback((data) => {
```
Comment (silva-fj) [inline src/screen/CreateAccount/CreateAccountScreen.tsx:31]: This should be encapsulated within the `AccountsContext`.

**Review by silva-fj — COMMENTED** [review]
Great work @ishrak-imam, let's keep this branch around in case we need to fallback to this solution later.

You can add a TODO to add the ability to sign transactions getting the keyringPair from the keyring and also add comments on the things you know we would need to improve so we can remember

File: src/context/AccountsContext.tsx (hunk 1 above)
Comment (ishrak-imam) [reply src/context/AccountsContext.tsx:133]: It takes some time to load the html into the webview, before that the webviewref is not ready.

Could have been better if it is possible to listen for the ref change.

**Pull request closed without merging**